vercel/data_source_project_rolling_release.go

```diff
@@ -0,0 +1,234 @@
+				Computed:            true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"target_percentage": schema.Int64Attribute{
+							MarkdownDescription: "The percentage of traffic to route to this stage.",
+							Computed:            true,
+						},
+						"duration": schema.Int64Attribute{
+							MarkdownDescription: "The duration in minutes to wait before advancing to the next stage. Present for automatic advancement type.",
+							Computed:            true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// ProjectRollingReleaseDataSourceModel reflects the structure of the data source.
+type ProjectRollingReleaseDataSourceModel struct {
+	AdvancementType types.String `tfsdk:"advancement_type"`
+	Stages          types.List   `tfsdk:"stages"`
+	ProjectID       types.String `tfsdk:"project_id"`
+	TeamID          types.String `tfsdk:"team_id"`
+}
+
+func (d *projectRollingReleaseDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data ProjectRollingReleaseDataSourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var out client.RollingReleaseInfo
+	var err error
+	var convertedData ProjectRollingReleaseDataSourceModel
+	var diags diag.Diagnostics
+
+	// Retry up to 10 times (10s) for eventual consistency
+	for i := 0; i < 10; i++ {
+		out, err = d.client.GetRollingRelease(ctx, data.ProjectID.ValueString(), data.TeamID.ValueString())
+		if err == nil && out.RollingRelease.Enabled && len(out.RollingRelease.Stages) > 0 {
+			break // found config
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	if client.NotFound(err) {
+		resp.Diagnostics.AddError(
+			"Error reading project rolling release",
+			fmt.Sprintf("No project rolling release found with id %s %s", data.TeamID.ValueString(), data.ProjectID.ValueString()),
+		)
+		return
+	}
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading project rolling release",
+			fmt.Sprintf("Could not get project rolling release %s %s, unexpected error: %s",
+				data.TeamID.ValueString(),
+				data.ProjectID.ValueString(),
+				err,
+			),
+		)
+		return
+	}
+
+	convertedData, diags = convertResponseToRollingReleaseDataSourceWithConfig(out, data, ctx)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Info(ctx, "converted data before setting state", map[string]any{
+		"project_id":       convertedData.ProjectID.ValueString(),
+		"team_id":          convertedData.TeamID.ValueString(),
+		"advancement_type": convertedData.AdvancementType.ValueString(),
+		"stages_is_null":   convertedData.Stages.IsNull(),
+		"stages_length":    len(convertedData.Stages.Elements()),
+	})
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &convertedData)...)
+}
+
+// Like convertResponseToRollingReleaseDataSource, but if API is empty/disabled, use config if present
```

Comment:
Comment is not needed?

Reply:
removed in a8847abd5eb0a6a42ff17dd9036c8dfb904d999c